contracts/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,167 @@
+///
+access(all) contract RandomBeaconHistory {
+
+    /// The height at which the first source of randomness was recorded
+    access(contract) var lowestHeight: UInt64?
+    /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
+    access(contract) let randomSourceHistory: [[UInt8]]
+
+    /// The path of the Heartbeat resource in the deployment account
+    access(all) let HeartbeatStoragePath: StoragePath
+
+    /* --- Hearbeat --- */
+    //
+    /// The Heartbeat resource containing each block's source of randomness in sequence
+    ///
+    access(all) resource Heartbeat {
+
+        /// Callable by owner of the Heartbeat resource, Flow Service Account, records the provided random source
+        ///
+        /// @param randomSourceHistory The random source to record
+        ///
+        access(all) fun heartbeat(randomSourceHistory: [UInt8]) {
+
+            let currentBlockHeight = getCurrentBlock().height
+            if RandomBeaconHistory.lowestHeight == nil {
+                RandomBeaconHistory.lowestHeight = currentBlockHeight
+            }
+
+            RandomBeaconHistory.randomSourceHistory.append(randomSourceHistory)
+        }
+    }
+
+    /* --- RandomSourceHistory --- */
+    //
+    /// Represents a random source value for a given block height
+    ///
+    access(all) struct RandomSource {
+        access(all) let blockHeight: UInt64
+        access(all) let value: [UInt8]
+    
+        init(blockHeight: UInt64, value: [UInt8]) {
+            self.blockHeight = blockHeight
+            self.value = value
+        }
+    }
+
+    /* --- RandomSourceHistoryPage --- */
+    //
+    /// Contains RandomSource values ordered chronologically according to associated block height
+    ///
+    access(all) struct RandomSourceHistoryPage {
+        access(all) let page: UInt64
+        access(all) let perPage: UInt64
+        access(all) let totalLength: UInt64
+        access(all) let values: [RandomSource]
+    
+        init(page: UInt64, perPage: UInt64, totalLength: UInt64, values: [RandomSource]) {
+            self.page = page
+            self.perPage = perPage
+            self.totalLength = totalLength
+            self.values = values
+        }
+    }
+
+    /* --- Contract Methods --- */
+    //
+    /// Getter for the source of randomness at a given block height. Panics if the requested block height either
+    /// precedes or exceeds the recorded history. Note that a source of randomness for block n will not be accessible
+    /// until block n+1.
+    ///
+    /// @param atBlockHeight The block height at which to retrieve the source of randomness
+    ///
+    /// @return The source of randomness at the given block height as RandomSource struct
+    ///
+    access(all) fun sourceOfRandomness(atBlockHeight: UInt64): RandomSource {
+        pre {
+            self.lowestHeight != nil: "History has not yet been initialized"
```

Comment:
Side note: It would be great to have a way to have a way of optional-binding in conditions, e.g. be able to say `let lowestHeight = self.lowestHeight` here instead, like in an `if let`, and then use `lowestHeight` instead of the repeated `self.lowestHeight!` in the condition below and in the body